### PR TITLE
Two Phase commit

### LINKED ISSUE #1016: [2-phase commit] pre commit directory
Upload could be rolled back with delete, rename could be rolled back with re-rename to old name, etc.
The two phase commit will introduce pre-commit directory and pre-commit table.
Precommit-directory --> To store file that handles upload and update.
Precommit-table --> To handle move, rename, copy, move, directory creation and delete. We might go with adding a bool column in existing table that specify if a file path is precommit or not.

### LINKED ISSUE #1038: [2-phase commit] rollback blobber
implement blobber rollback according to specification

### PR DESCRIPTION
**Pull request by Hitenjain14**
### Changes

Added preCommitDir and rollback handler

### Associated PRs (Link as appropriate):
- 0chain: https://github.com/0chain/0chain/pull/2340
- gosdk: https://github.com/0chain/gosdk/pull/934
- system_test: 
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | two-commit-testing  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | two-commit  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | two-commit-testing  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | two-commit  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | two-commit-testing  |
| zwalletcli  | staging  |
| 0chain  | pr-2340-a43aa273  |
| blobber  | two-commit  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | two-commit-testing  |
| zwalletcli  | staging  |
| 0chain  | rollback-wm  |
| blobber  | two-commit  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | rollback-wm  |
| blobber  | two-commit  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | rollback-wm  |
| blobber  | two-commit  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | two-commit-testing  |
| zwalletcli  | staging  |
| 0chain  | rollback-wm  |
| blobber  | two-commit  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by Hitenjain14** [comment]
Recent run https://github.com/0chain/system_test/actions/runs/4886059001

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | two-commit-testing  |
| zwalletcli  | staging  |
| 0chain  | rollback-wm  |
| blobber  | two-commit  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | two-commit-testing  |
| zwalletcli  | staging  |
| 0chain  | rollback-wm  |
| blobber  | two-commit  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

File: code/go/0chain.net/blobbercore/allocation/allocationchange.go [hunk 1]
```diff
@@ -118,6 +122,24 @@ func GetAllocationChanges(ctx context.Context, connectionID, allocationID, clien
...
 }
 
+// Get the precommit changes for the allocation
+func GetAllocationPreCommitChanges(ctx context.Context, allocationId, clientId string) (*AllocationChangeCollector, error) {
+
+	cc := &AllocationChangeCollector{}
+
+	db := datastore.GetStore().GetTransaction(ctx)
+
+	err := db.Where("allocation_id = ? and client_id = ? and is_precommit = ?", allocationId, clientId, true).Preload("Changes").First(cc).Error
```
Comment (dabasov) [inline code/go/0chain.net/blobbercore/allocation/allocationchange.go]: is client_id needed here?

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | two-commit-testing  |
| zboxcli  | two-commit-testing  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | two-commit  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | two-commit-testing  |
| zboxcli  | two-commit-testing  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | two-commit  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | two-commit-testing  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | two-commit  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Review by dabasov — APPROVED** [review]
lgtm

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | two-commit-testing  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Pull request merged**